### PR TITLE
wasmtime: Extract cranelift/lightbeam compilers to separate crates

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit extracts the two implementations of `Compiler` into two
separate crates, `wasmtime-cranelfit` and `wasmtime-lightbeam`. The
`wasmtime-jit` crate then depends on these two and instantiates them
appropriately. The goal here is to start reducing the weight of the
`wasmtime-environ` crate, which currently serves as a common set of
types between all `wasmtime-*` crates. Long-term I'd like to remove the
dependency on Cranelift from `wasmtime-environ`, but that's going to
take a lot more work.

In the meantime I figure it's a good way to get started by separating
out the lightbeam/cranelift function compilers from the
`wasmtime-environ` crate. We can continue to iterate on moving things
out in the future, too.

### CODE REVIEW

File: crates/lightbeam/wasmtime/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+//!
+//! This crates provides an implementation of [`Compiler`] in the form of
+//! [`Lightbeam`].
+
+use cranelift_codegen::binemit;
+use cranelift_codegen::ir::{self, ExternalName};
+use cranelift_codegen::isa;
+use cranelift_wasm::{
+    DefinedFuncIndex, DefinedGlobalIndex, DefinedMemoryIndex, DefinedTableIndex, FuncIndex,
+    GlobalIndex, MemoryIndex, SignatureIndex, TableIndex,
```
Comment (yurydelendik) [inline crates/lightbeam/wasmtime/src/lib.rs:11]: It is possible to move some of this imports into "crates/environ/src/data_structures.rs"?

Comment (alexcrichton) [reply crates/lightbeam/wasmtime/src/lib.rs:12]: 👍

Comment (yurydelendik) [reply crates/lightbeam/wasmtime/src/lib.rs:12]: Nice. Sorry, I also meant other no `use cranelift_codegen::XXX` above -- they can be re-exported in wasmtime_environ too thus removing direct dependency on cranelift crates.

Comment (alexcrichton) [reply crates/lightbeam/wasmtime/src/lib.rs:12]: I was thinking of doing that yeah but I realized that lightbeam itself depends on `cranelift-codegen`, so I figured it was probably best to just do the same thing.

**Pull request merged**